lib/auth/oidc/client_assertion.go

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package oidc
+
+import (
+	"crypto/rsa"
+	"crypto/sha1"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"os"
+	"time"
+
+	gojwt "github.com/golang-jwt/jwt/v5"
+	cass "github.com/hashicorp/cap/oidc/clientassertion"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+func BuildClientAssertionJWT(config *structs.ACLAuthMethodConfig, nomadKey *rsa.PrivateKey, nomadKID string) (*cass.JWT, error) {
+	// should already be validated by caller, but just in case.
+	if config == nil || config.OIDCClientAssertion == nil {
+		return nil, errors.New("no auth method config or client assertion")
+	}
+
+	// this is all we use config for
+	clientID := config.OIDCClientID
+	// client assertion-specific info is in here
+	as := config.OIDCClientAssertion
+
+	// this should have also happened long before, but again, just in case.
+	if err := as.Validate(); err != nil {
+		return nil, err
+	}
+
+	opts := []cass.Option{
+		cass.WithHeaders(as.ExtraHeaders),
+	}
+
+	switch as.KeySource {
+
+	case structs.OIDCKeySourceClientSecret:
+		algo := cass.HSAlgorithm(as.KeyAlgorithm)
+		return cass.NewJWTWithHMAC(clientID, as.Audience, algo, as.ClientSecret, opts...)
+
+	case structs.OIDCKeySourceNomad:
+		opts = append(opts, cass.WithKeyID(nomadKID))
+		return cass.NewJWTWithRSAKey(clientID, as.Audience, cass.RS256, nomadKey, opts...)
+
+	case structs.OIDCKeySourcePrivateKey:
+		algo := cass.RSAlgorithm(as.KeyAlgorithm)
+		rsaKey, err := getCassPrivateKey(as.PrivateKey)
+		if err != nil {
+			return nil, err
+		}
+		keyID := as.PrivateKey.KeyID
+		if keyID == "" {
+			cert, err := getCassCert(as.PrivateKey)
+			if err != nil {
+				return nil, err
+			}
+			keyID = X5T(cert)
+		}
+		opts = append(opts,
+			cass.WithKeyID(keyID),
+		)
+		return cass.NewJWTWithRSAKey(clientID, as.Audience, algo, rsaKey, opts...)
+
+	default: // this shouldn't happen, but just in case
+		return nil, fmt.Errorf("unknown OIDC KeySource %q", as.KeySource)
+	}
+}
+
+// getCassPrivateKey parses the structs.OIDCClientAssertionKey PemKeyFile
+// or PemKey, depending on which is set.
+func getCassPrivateKey(k *structs.OIDCClientAssertionKey) (key *rsa.PrivateKey, err error) {
+	var bts []byte
+	var source string // for informative error messages
+
+	// pem file on disk
+	if k.PemKeyFile != "" {
+		source = "PemKeyFile"
+		bts, err = os.ReadFile(k.PemKeyFile)
```

Comment:
While allowing users to specify arbitrary files to be read on servers scares me, the file is always specified by admin's who can presumably do much worse.

An interesting side effect of using PemKeyFile over PemKey is that PemKeyFile can be rotated without interacting with Nomad at all since we read the file on every login. The con being that users must ensure new/recovered servers have the files before they're able to process SSO logins.

Probably a good tradeoff to call out in docs.

Reply:
@gulducat and I talked about this and have some tests on that too https://github.com/hashicorp/nomad/compare/f-oidc-client-assertion...oidc/security-tests?expand=1#diff-b0cf3a779250f5e1017d77ae352865bdc5d393d7f93d76a0e1785629370c7b6cR340-R373 (needs some rebase)😄 Here is the [slack](https://hashicorp.slack.com/archives/C05EJN96VNH/p1741363337512599?thread_ts=1741013900.496149&cid=C05EJN96VNH) 🧵